eq3bt/connection.py

```diff
@@ -15,46 +17,83 @@
 class BTLEConnection(btle.DefaultDelegate):
     """Representation of a BTLE Connection."""
 
-    def __init__(self, mac):
+    def __init__(self, mac, keep_connected=False):
         """Initialize the connection."""
         btle.DefaultDelegate.__init__(self)
 
+        self._ifaces = self.get_hci_ifaces()
+        self._iface_idx = 0
+
         self._conn = None
         self._mac = mac
         self._callbacks = {}
+        self._keep_connected = keep_connected
+    
+    def next_iface(self):
+        self._nr_conn_errors += 1
+        self._iface_idx = (self._iface_idx + 1) % len(self._ifaces)
+        if self._nr_conn_errors >= len(self._ifaces)*2:
+            return False
+        return True
+
+    def get_hci_ifaces(self):
+        iface_list = []
+        bus = dbus.SystemBus()
+        manager = dbus.Interface(bus.get_object("org.bluez", "/"), "org.freedesktop.DBus.ObjectManager")
+        objects = manager.GetManagedObjects()
+        for path, interfaces in objects.items():
+            adapter = interfaces.get("org.bluez.Adapter1")
+            if adapter is None:
+                continue
+            iface_list.append(re.search(r'\d+$', path)[0])
+        return iface_list
```

Comment:
This would introduce hard dependency for dbus (which may or may not be available on all environments), so it's a no go. Could you please create a separate PR for adding optional support for multiple interfaces?

Also, are you sure bluepy does not provide an interface for enumerating over existing adapters? That would be the preferred way over manually scraping bluez responses.

Reply:
Unfortunately bluepy doesn't provide such a functionality, that was the first place I searched for. I could find this one from bluez which is the functionality I implemented:
https://github.com/pauloborges/bluez/blob/master/test/test-adapter

Another possible solution would be to remove this one from eq3bt and add an additional parameter "iface".

Reply:
Hmm, I think the best solution is to refactor the connection handling (so that connect can be done separately outside the `__enter__`) and expose `iface` parameter for it as you propose. The eq3bt cli tool could also optionally expose `--interface` option, and the downstream users can use whatever they want (like dbus) to locate available interfaces.

Reply:
Just to be sure that we are on the same page:
 - add `connect(iface)` method
 - add `iface` parameter to `Thermostat`
 - remove dbus and get adapters completely

Reply:
Sounds good to me!

* `connect(iface)` will be useful for your use case, where you want to fall back to other interfaces if the connection fails
* Adding optional `iface` to `Thermostat` allows forcing to use specific controller (e.g., via cli) during the initialization phase, in case someone wants to do that.

The real fix for obtaining interfaces should be done inside `bluepy`, but I think the library is not very actively maintained, notwithstanding the problem with dbus availability (iirc there was also some problems getting that library easily pip-installable due to some compiling needed during the setup process). Last time I did some dbus interfacing, I used python-dbus-next which is pure python only library.